### ISSUE TITLE
IBeacon decoder index out of bounds

### ISSUE THREAD

**Issue by markoer** [eclipse/kura#1863]
If by any chance this check:

https://github.com/eclipse/kura/blob/961029abd4c86aa6c738f2b2aaf733d1c3bdedbc/kura/org.eclipse.kura.ble.ibeacon.provider/src/main/java/org/eclipse/kura/internal/ble/ibeacon/BluetoothLeIBeaconDecoderImpl.java#L68

succeeds on the first pass, an ArrayIndexOutOfBoundsException (-1) is bound to be thrown here:

https://github.com/eclipse/kura/blob/961029abd4c86aa6c738f2b2aaf733d1c3bdedbc/kura/org.eclipse.kura.ble.ibeacon.provider/src/main/java/org/eclipse/kura/internal/ble/ibeacon/BluetoothLeIBeaconDecoderImpl.java#L73

I guess this could be a problem.

The other thing is that the Decoder cannot decode Encoder's output. This may be OK, if they were not designed to be used one for the other.
The good thing is that somehow it all works...

As before, if I'm missing something, close the issue freely.

**Comment by pierantoniomerlino** [comment]
@markoer The iBeacon advertisement packet format is as follows:

02 01 1a 1a ff 4c 00 02 15 aa aa aa aa bb bb cc cc dd dd ee ee ee ee ee 00 00 00 00 00 

where 
02 Data length
01 Data type
1a LE and BR/EDR flags
1a Data length 
ff Data type
4c 00 02 15 manufacturer code
aa aa aa aa bb bb cc cc dd dd ee ee ee ee ee UUID
00 00 major
00 00 minor
00 tx power

So, the BluetoothLeIBeaconDecoderImpl (line 68) tries to detect the DataType (ff) and the iBeacon prefix (4c 00 02 15). These data are not at the beginning of the packet, since the previous data contain the LE flags that the code parses at line 73-77. However, I agree with you that some check on the pointer position is needed.

**Comment by pierantoniomerlino** [comment]
@markoer Why you said that the decoder cannot decode the encoder's output?

**Comment by markoer** [comment]
@pierantoniomerlino Try this test to see what I meant:
https://github.com/eclipse/kura/blob/351ae0fbc8a1dc8dc2f7bcbe313dd52d9e27e617/kura/test/org.eclipse.kura.internal.ble.ibeacon.test/src/test/java/org/eclipse/kura/internal/ble/ibeacon/BluetoothLeIBeaconEncoderDecoderImplTest.java#L46

**Comment by pierantoniomerlino** [comment]
@markoer Ahh! The encoder and decoder are not symmetric indeed. 

During the encoding process, the iBeacon is processed to obtain a byte stream that can be directly passed to the hcitool command. So, it doesn't contain only the reports but also the header.

During the decoding process, the byte stream is filtered by BluetoothLeUtil.parseLEAdvertisement that extracts the advertisingReportRecords and rssi from the BLE Advertising packet. Then the reports are decoded.

Long story short, the test works if the rssi is 0 and we remove the first byte from the array passed to the decoder:
```java
beacon.setRssi(0);
...
BluetoothLeIBeacon decoded = decoder.decode(Arrays.copyOfRange(encoded, 1, encoded.length));
```

I can do a PR for that, if you agree.

**Comment by markoer** [comment]
Of course, go ahead :)
Thanks for the explanation(s).
